### PR TITLE
ci: Automated tests using GitHub Actions

### PR DESCRIPTION
**Pull request by teolemon**
Automated tests using GitHub Actions

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -5,22 +5,36 @@ on:
...
-          path: ~/.composer/cache/files
-          key: dependencies-composer-${{ hashFiles('composer.json') }}
       - name: Setup PHP
         uses: shivammathur/setup-php@v2
         with:
-          php-version: 7.3
-          extensions: dom, curl, libxml, mbstring, zip, pcntl, pdo, sqlite, pdo_sqlite
-          coverage: none
+          php-version: ${{ matrix.php }}
+          extensions: dom, curl, libxml, mbstring, zip, pcntl, pdo, sqlite, pdo_sqlite, xdebug
```
Comment (epalmans) [inline .github/workflows/ci.yml]: Not running code coverage analysis? Then: remove xdebug extension

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -5,22 +5,36 @@ on:
...
+          coverage: xdebug
```
Comment (epalmans) [inline .github/workflows/ci.yml]: Not running code coverage analysis? Then: remove this line

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -5,22 +5,36 @@ on:
...
+        with:
+          path: ${{ steps.composer-cache.outputs.dir }}
+          key: ${{ matrix.php }}-composer-${{ hashFiles('**/composer.lock') }}
+          restore-keys: ${{ matrix.php }}-composer-
       - name: Install Composer dependencies
-        run: composer install --prefer-dist --no-interaction --no-suggest
+        run: composer install --prefer-dist --no-interaction --no-progress --no-suggest
       - name: Execute tests
-        run: vendor/bin/phpunit --verbose
+        run: vendor/bin/phpunit --verbose --coverage-clover ./build/coverage.clover
```
Comment (epalmans) [inline .github/workflows/ci.yml]: Not running code coverage analysis? Then: don't bother generating a coverage-clover file

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -5,22 +5,36 @@ on:
...
+      - name: Downloading scrutinizer ocular.phar
+        run: wget https://scrutinizer-ci.com/ocular.phar
+      - name: Uploading code coverage to scrutinize
+        run: php ocular.phar code-coverage:upload --format=php-clover ./build/coverage.clover
```
Comment (epalmans) [inline .github/workflows/ci.yml]: Not running code coverage analysis through Scrutinizer? Then: don't bother uploading the coverage.clover file

**Review by epalmans — CHANGES_REQUESTED** [review]
Few suggestions on cleaning up things when not planning to add code coverage analysis on this repo - see comments

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        php: ['7.2', '7.3', '7.4']
+    name: Tests - PHP ${{ matrix.php }} 
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2
+      - name: Setup PHP
+        uses: shivammathur/setup-php@v2
+        with:
+          php-version: ${{ matrix.php }}
+          extensions: dom, curl, libxml, mbstring, zip, pcntl, pdo, sqlite, pdo_sqlite, xdebug
```
Comment (teolemon) [inline .github/workflows/ci.yml]:
```suggestion
          extensions: dom, curl, libxml, mbstring, zip, pcntl, pdo, sqlite, pdo_sqlite
```

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+          coverage: xdebug
```
Comment (teolemon) [inline .github/workflows/ci.yml]:
```suggestion
```

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+      - name: Cache dependencies
+        uses: actions/cache@v1
+        with:
+          path: ${{ steps.composer-cache.outputs.dir }}
+          key: ${{ matrix.php }}-composer-${{ hashFiles('**/composer.lock') }}
+          restore-keys: ${{ matrix.php }}-composer-
+      - name: Install Composer dependencies
+        run: composer install --prefer-dist --no-interaction --no-progress --no-suggest
+      - name: Execute tests
+        run: vendor/bin/phpunit --verbose --coverage-clover ./build/coverage.clover
```
Comment (teolemon) [inline .github/workflows/ci.yml]:
```suggestion
```

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+      - name: Downloading scrutinizer ocular.phar
+        run: wget https://scrutinizer-ci.com/ocular.phar
+      - name: Uploading code coverage to scrutinize
+        run: php ocular.phar code-coverage:upload --format=php-clover ./build/coverage.clover
```
Comment (teolemon) [inline .github/workflows/ci.yml]:
```suggestion
```

File: .github/workflows/ci.yml (hunk 3 above)
Comment (teolemon) [reply .github/workflows/ci.yml]: thanks

File: .github/workflows/ci.yml (hunk 4 above)
Comment (teolemon) [reply .github/workflows/ci.yml]: thanks

File: .github/workflows/ci.yml (hunk 1 above)
Comment (teolemon) [reply .github/workflows/ci.yml]: thanks

**Review by epalmans — APPROVED** [review]

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+        run: echo "::set-output name=dir::$(composer config cache-files-dir)"
+      - name: Cache dependencies
+        uses: actions/cache@v1
+        with:
+          path: ${{ steps.composer-cache.outputs.dir }}
+          key: ${{ matrix.php }}-composer-${{ hashFiles('**/composer.lock') }}
+          restore-keys: ${{ matrix.php }}-composer-
+      - name: Install Composer dependencies
+        run: composer install --prefer-dist --no-interaction --no-progress --no-suggest
+      - name: Execute tests
```
Comment (Dwarfex) [inline .github/workflows/ci.yml]:
```suggestion
      - name: Execute tests
        run: vendor/bin/phpunit --verbose
```

But we should still execute the tests ;)

**Review by Dwarfex — CHANGES_REQUESTED** [review]
We should still run the tests ;)

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+on:
+  push:
+  pull_request:
+jobs:
+  tests:
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        php: ['7.2', '7.3', '7.4']
```
Comment (Dwarfex) [inline .github/workflows/ci.yml]:
```suggestion
        php: ['7.3', '7.4', '8.0']
```
I don't know if this already works with ubuntu-latest, but we should try to run the tests on PHP8.0 and drop 7.2 as it is already EoL

File: .github/workflows/ci.yml (hunk 9 above)
Comment (epalmans) [reply .github/workflows/ci.yml]: oops, good catch

**Comment by Dwarfex** [comment]
Will close this - is integrated into a new 03.x-dev Branch

**Pull request closed without merging**